### PR TITLE
Load app description and developer notes from readme files

### PR DESCRIPTION
**Pull request by jdidion**
This PR introduces the concept of "adjunct" files, which are files in the same directory as the WDL that provide additional information.

Currently, there are two types of adjunct files:

* Readme: must have name `Readme.<wdlname>.<taskname>.md`, where wdlname is the name of the WDL file without the ".wdl" extension, and taskname is the name of the task. Filenames are matched case-insensitively.
* Developer Notes: must have name `Readme.developer.<wdlname>.<taskname>.md`

During parsing, adjunct files are discovered by name. During conversion to IR, if the task metadata does not have "description" and there was a Readme adjunct discovered, the description is set to the contents of that file. Similarly, if developer_notes is not defined in the applet meadata and Developer Notes adjunct was discovered, the developer_notes are set to the contents of that file.

Please review this file and comment on the approach I've taken. If there is a better way to do this, or a more appropriate way to organize the code, I want to know. I also realize that this currently only works for the main WDL file - adjuncts are not being discovered for imports. There is also not a great method of discovering remote adjuncts other than to try to fetch the file with the expected name and see if it exists.

### CODE REVIEW

File: src/main/scala/dxWDL/util/Adjunct.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  def findAdjunctFiles(path: Path): Map[String, Vector[AdjunctFile]] = {
+    val file = path.toFile
+    val parentDir = file.getParentFile
+    var wdlName = file.getName
+
+    if (wdlName.endsWith(".wdl")) {
+      wdlName = wdlName.dropRight(4)
+    }
+
+    val readmeRegexp = s"(?i)readme.${wdlName}.(.*)(?:.md)".r
```
Comment (sstadick) [inline src/main/scala/dxWDL/util/Adjunct.scala]: Perhaps this is intentional, but should the '.'s be escaped? currently the following matches:
```
val test = s"(?i)readme.GLnexus.(.*)(?:.md)".r
"readme#glnexus#stage1#md" match { case test(stage) => println(stage); case _ => println("None")}
```

Comment (sstadick) [reply src/main/scala/dxWDL/util/Adjunct.scala]: Should these be .+ instead of .*?
The (:?.md) could just be \.md right?

**Comment by sstadick** [comment]
Thinking about how to resolve imports, what if we made this more explicit? So if you want a description, you either have a string describing it, or you have a path to a readme with the description? And that path must be relative to the top level of your project (or something else)? This could also allow for maybe having all your docs in one readme, and then passing the path plus the header for your section.

EX:
Readme.md
```
# This is a workflow for doing cool stuff

## add

1 + 1 = 2

### Finer details about Stage1

1 + 1 + 1 = 3

## mul

2 * 2 = 4
```

```wdl
version 1.0

# A WDL workflow with expressions, but without branches and loops.

workflow wf_readme {
    input {
        Int x = 3
        Int y = 5
    }

    call add {
        input: a=x, b=y
    }

    Int z = add.result + 1

    call mul { input: a=z, b=5 }

    call inc { input: i= z + mul.result + 8}

    output {
        Int result = inc.result
    }
}


task add {
    input {
        Int a
        Int b
    }

    meta {
        developer_notes: "Developer notes defined in WDL"
        description: "Readme.md#add"
    }

    command {
        echo $((${a} + ${b}))
    }

    output {
        Int result = read_int(stdout())
    }
}

task mul {
    input {
        Int a
        Int b
    }

    meta {
        description: "Description defined in WDL"
    }

    command {
        echo $((a * b))
    }
    output {
        Int result = a * b
    }
}

task inc {
    input {
        Int i
    }

    command <<<
        python -c "print(${i} + 1)"
    >>>
    output {
        Int result = read_int(stdout())
    }
}
```

A lib like https://github.com/vsch/flexmark-java could be used to for parsing the markdown.

Granted this is a more complex approach, and what you have now works with the exception of imports.

File: src/main/scala/dxWDL/util/Adjunct.scala (hunk 1 above)
Comment (jdidion) [reply src/main/scala/dxWDL/util/Adjunct.scala]: Yep, correct on both counts. Fixed now.

**Comment by jdidion** [comment]
Regarding your first point, I thought about allowing a path/URL, but it seems overly complicated. We could add it if people ask for it. I would want to then allow `description` to be either a string or an object with either a `path` or `url` key; i.e. I don't want to have to try to figure out from a string whether it is a path, url, or text description.

Regarding your second point, this is a good suggestion but there is a separate ticket for it. There are many potential gotchas with trying to extract different parts of a markdown file. I think it is lower priority unless users really clamor for it.

**Comment by sstadick** [comment]
Makes sense! This method works and doesn't preclude something more complex being added on top of it at a future time.

File: src/main/scala/dxWDL/base/Utils.scala [hunk 2]
```diff
@@ -102,6 +102,11 @@ object Utils {
     sb.toString
   }
 
+  // This one seems more idiomatic
+  def readFileContent2(path: Path): String = {
```
Comment (orodeh) [inline src/main/scala/dxWDL/base/Utils.scala]: Sure, just replace the old implementation then. This isn't used in performance contexts anyway.

File: src/main/scala/dxWDL/compiler/IR.scala [hunk 3]
```diff
@@ -22,8 +22,21 @@ object IR {
   val REORG = "reorg"
   val CUSTOM_REORG_CONFIG = "reorg_config"
 
+  // The following keywords/types correspond to app(let) attributes from dxapp.json (except for
+  // inputSpec/outputSpec attributes, which are defined separately). These attributes can be used
+  // in the meta section of task WDL, and will be parsed out and used when generating the native
+  // app.
+  val META_DESCRIPTION = "description"
+  val META_DEVELOPER_NOTES = "developer_notes"
+  val META_TYPE = "type"
```
Comment (orodeh) [inline src/main/scala/dxWDL/compiler/IR.scala]: Can you make sure this doesn't conflict with with what DxNI generates? You can (and should) run the regression tests for this (staging, ./scripts/run_tests.py).

**Comment by jtratner** [comment]
Could (instead) the file be in the same directory as the WDL file.

I.e., if I have layout like:

```
./batch.wdl
./lane.wdl
./tools/vcf.wdl
./tools/bam.wdl
```

then I have

```
./Readme.batch.sometask.md
./Readme.lane.sometask.md
./tools/Readme.vcf.sometask.md
./tools/Readme.bam.sometask.md
```

etc, and those would be picked up appropriately.

Also nice would be something like `dxwdl make-readmes` which would then create templates for all of the README files above, with appropriate formatting so it renders correctly on the platform.

**Comment by jdidion** [comment]
@jtratner yes that's exactly the way it works. The make-readmes command is a great idea - I'll add a separate issue for that.

File: src/main/scala/dxWDL/base/Utils.scala (hunk 2 above)
Comment (jdidion) [reply src/main/scala/dxWDL/base/Utils.scala]: I will eventually, but readFileContent is used in a lot of places, so I'll do it in a separate PR.

**Comment by jdidion** [comment]
It now works for all tasks. I will add do workflow adjuncts in a separate PR.

**Pull request merged**